[Guide] ES6: Compare var and let scopes. Fixes and enhancements. (#34636)
freeCodeCamp/freeCodeCamp: guide/english/certifications/javascript-algorithms-and-data-structures/es6/compare-scopes-of-the-var-and-let-keywords/index.md

---
title: Compare Scopes of the var and let Keywords
---

![:triangular_flag_on_post:](https://forum.freecodecamp.com/images/emoji/emoji_one/triangular_flag_on_post.png?v=3 ":triangular_flag_on_post:") Remember to use <a>**`Read-Search-Ask`**</a> if you get stuck. Try to pair program ![:busts_in_silhouette:](https://forum.freecodecamp.com/images/emoji/emoji_one/busts_in_silhouette.png?v=3 ":busts_in_silhouette:") and write your own code ![:pencil:](https://forum.freecodecamp.com/images/emoji/emoji_one/pencil.png?v=3 ":pencil:")

## Problem Explanation:

> _Fix the code so that `i` declared in the if statement is a separate variable than `i` declared in the first line of the function._

Reviewer: Continuing from @Manish-Giri , this is more confusing than ever. What do you mean by separate variable? Running on separate machines? If you really really want to do this, how about:
```suggestion
Change the code so that the variable `i` declared in the if block is separately scoped than the variable `i` declared at the beginning of the function
```

Author: Those changes look perfect in my opinion. I didn't mean to confuse anyone, as I said I was trying to avoid giving users the solution in the instructions, I couldn't think of a better way to explain it and instead cited the original exercise instructions: 
![image](https://user-images.githubusercontent.com/12052430/54948646-74a3c500-4f3d-11e9-98ad-3169b7b89c34.png)
If it's confusing maybe we could change those as well.

Reviewer: @RandellDawson @moT01 What do you think?

Reviewer 2: yes, sure @thecodingaviator @AdrianSkar, sounds good to me - someone make the commit please and we can give it a final review/merge

Reviewer 3: I tried to commit, but I get the message "You do not have permission to push to this repository. " for some reason.  @AdrianSkar I am fine with @thecodingaviator's suggestion.  Please commit it.  Thanks.

---
title: Compare Scopes of the var and let Keywords
---

![:triangular_flag_on_post:](https://forum.freecodecamp.com/images/emoji/emoji_one/triangular_flag_on_post.png?v=3 ":triangular_flag_on_post:") Remember to use <a>**`Read-Search-Ask`**</a> if you get stuck. Try to pair program ![:busts_in_silhouette:](https://forum.freecodecamp.com/images/emoji/emoji_one/busts_in_silhouette.png?v=3 ":busts_in_silhouette:") and write your own code ![:pencil:](https://forum.freecodecamp.com/images/emoji/emoji_one/pencil.png?v=3 ":pencil:")

> ### Problem explanation:

Reviewer 4: ```suggestion
## Problem Explanation:
```
Please revert `Explanation` as @RandellDawson mentioned, and increase the heading level to h2 `##`.

_Fix the code so that `i` declared in the if statement is a separate variable than `i` declared in the first line of the function._

## ![:speech_balloon:](https://forum.freecodecamp.com/images/emoji/emoji_one/speech_balloon.png?v=3 ":speech_balloon:") Hint: 1

> *  _Be certain not to use the `var` keyword anywhere in your code._

Author: I disagree. The original text gives away the solution so its not actually a hint.

> *  _Be certain not to use the `var` keyword anywhere in your code._

Reviewer 4: > _Fix the code so that `i` declared in the if statement is a separate variable than `i` declared in the first line of the function._

@AdrianSkar This statement is not really clear. One could argue that since both variables are named `i`, they're the same variable, reused in a different scope. To avoid any confusion, please revert to the original sentence.

> *  _Be certain not to use the `var` keyword anywhere in your code._

Author: Hi @Manish-Giri! I changed that text because I felt it "gives away" the solution but I didn't write the replacement, I used the [original exercise instructions](https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/es6/compare-scopes-of-the-var-and-let-keywords) (that's why I put it in italics). If you disagree we can just let it as it was, from my _newb_ perspective I like it better as in previous guide pages.

> *  _Be certain not to use the `var` keyword anywhere in your code._

Reviewer: I agree with @AdrianSkar on this topic, the original text gives away the solution. What I propose is that we let the first hint be as has been suggested by the OP, but with the following formatting modifications:
```suggestion
Be certain not to use the `var` keyword anywhere in your code since variables declared with var are globally scoped.
```

Then we add a second hint with the text:
```text
Find `var` and replace with `let`.
```

> *  _Be certain not to use the `var` keyword anywhere in your code._

Author: That's ok in my opinion. I put them in italics because it's a citation from the original instructions but that looks good too. Should I commit this changes?
Regarding the second hint it's actually the solution so not sure about that one. In any case I'm the inexperienced one so whatever you think it's best/could help someone with this lesson.

> *  _Be certain not to use the `var` keyword anywhere in your code._

Reviewer: @AdrianSkar Before committing it, I'd say we get the approval of either of @RandellDawson / @moT01 / @Manish-Giri

> _try to solve the problem now_

> *   Remember that `let`'s scope is limited to the block, function or statement in which you declare it.

Reviewer 4: ```suggestion
*   _Remember that `let`'s scope is limited to the block, function or statement in which you declare it._
```

- Make the hint text italicized, following the same formatting as your first hint
- Add a "Hint 2" sub heading just above this hint (you can use the same formatting as Hint 1)

> _try to solve the problem now_

> ## Spoiler alert!

Reviewer 4: ```suggestion
## Spoiler Alert!
```

    return i;
    }
```
![:rocket:](https://forum.freecodecamp.com/images/emoji/emoji_one/rocket.png?v=3 ":rocket:") [Run code at codepen.io](https://codepen.io/dylantyates/pen/wxwxRd)

> # Code Explanation:

Reviewer 4: ```suggestion
## Code Explanation:
```

> ![:rocket:](https://forum.freecodecamp.com/images/emoji/emoji_one/rocket.png?v=3 ":rocket:") [Run code at codepen.io](https://codepen.io/dylantyates/pen/wxwxRd)

Reviewer: ```suggestion

```

Since we are in the process of phasing out links containing user controlled code, I'd appreciate if you remove this line
